drivers/adc/CMakeLists.txt

```diff
@@ -1,5 +1,10 @@
 zephyr_library()
 
+zephyr_library_compile_options_ifdef(CONFIG_COVERAGE_GCOV
```

Comment:
Are we really going to add this to EVERY cmake file? Why not just put it in the top level cmake?

Reply:
Unfortunately this is a limitation by CMAKE. we have to do it like this.

Reply:
@AdithyaBaglody Please look here: https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/posix/CMakeLists.txt#L12

Reply:
@aescolar it will enable coverage for all the files that will be compiled in the build. This will include all the files that come from ext,it will also include a lot of test case code.

Also I had done this in my first set of patches. it turns out that I need a lot more ram just to support this. Unfortunately Cmake is not matting things easier. If I have to specifically add a compiler flag for a single file, I have to make it into a library and then add the required compiler options/flags. This is a limitation on the CMake. 
I want this code to support as many devices as possible. But adding coverage for all files will restrict the code to be run on 1 or 2 devices at most.

Reply:
We have several options here.

Enable it by default, and then add -fno-coverage when we don't want it (opt-out).

```
zephyr_compile_options_ifdef(CONFIG_COVERAGE_GCOV ${COVERAGE_FLAGS})

zephyr_library_compile_options_ifdef(CONFIG_COVERAGE_GCOV ${NO_COVERAGE_FLAGS})
```

Use directory properties, but this can cause problems if the directory contains zephyr_sources calls as they are not affected by directory properties. Most kernel code does not contain zephyr_sources calls though, so this might work.

Reply:
@SebastianBoe I am a bit confused with this approach. isn't this as laborious as the current approach?. As in I have to go to all the folders that doesn't need this coverage and then add it. 

As you suggested i can add the directory properites to test folder. ( i havent tried this out but i saw that they are directly invoking target_sources which can be modified AFAIK)
For ext it is a bit more complicated, there are places where they call zephyr_sources. Which cant be changed with dir property.

Reply:
> I am a bit confused with this approach. isn't this as laborious as the current approach?. As in I have to go to all the folders that doesn't need this coverage and then add it.

I don't know the details, if there are many files that will need to be excluded then it is not a viable option.

Using directory properties will indeed require the banning of "zephyr_sources" in some directories.

Reply:
> I want this code to support as many devices as possible. But adding coverage for all files will restrict the code to be run on 1 or 2 devices at most.

I still do not understand why you can't put this in the top level drivers CMakefile and have to go and add it for each driver individually. What board configuration will work if we enable this on the top level?

Reply:
>I still do not understand why you can't put this in the top level drivers CMakefile

This is do with how cmake parses all the various CmakeList.txt files and figures out how to create the required bin. So in zephyr we have 3 ways to include source code. target_sources, zephyr_sources, zephyr_library_sources. We use these 3 in various places and the way to attach a simple flag to each vary significantly. 
As of now all the required files have been attached with the coverage option. If need be we can expand it later.
And I am guessing here, it will only run on qemu_x86 and mps2_an385. Last I tried in qemu_x86 I was getting a lot of stack corruption when I enabled by default. And I suggest we not enable this for the complete repo.